components/sensor/power_ina219/power_ina219.go

```diff
@@ -0,0 +1,229 @@
+	}
+	addr := attr.I2cAddr
+	if addr == 0 {
+		addr = defaultI2Caddr
+		logger.Infof("using i2c address : %d", defaultI2Caddr)
+	}
+
+	s := &ina219{
+		logger: logger,
+		bus:    i2cbus,
+		addr:   byte(addr),
+	}
+
+	err = s.calibrate()
+	if err != nil {
+		return nil, err
+	}
+
+	return s, nil
+}
+
+// ina219 is a i2c sensor device that reports voltage, current and power.
+type ina219 struct {
+	generic.Unimplemented
+	logger     golog.Logger
+	bus        board.I2C
+	addr       byte
+	currentLSB int64
+	powerLSB   int64
+	cal        uint16
+}
+
+type powerMonitor struct {
+	Shunt   int64
+	Voltage float64
+	Current float64
+	Power   float64
+}
+
+func (d *ina219) calibrate() error {
+	if senseResistor <= 0 {
+		return fmt.Errorf("ina219 calibrate: senseResistor value invalid %d", senseResistor)
+	}
+	if maxCurrent <= 0 {
+		return fmt.Errorf("ina219 calibrate: maxCurrent value invalid %d", maxCurrent)
+	}
+
+	d.currentLSB = maxCurrent / (1 << 15)
+	d.powerLSB = int64((maxCurrent*20 + (1 << 14)) / (1 << 15))
+	// Calibration Register = 0.04096 / (current LSB * Shunt Resistance)
+	// Where lsb is in Amps and resistance is in ohms.
+	// Calibration register is 16 bits.
+	cal := calibratescale / (d.currentLSB * int64(senseResistor))
+	if cal >= (1 << 16) {
+		return fmt.Errorf("ina219 calibrate: calibration register value invalid %d", cal)
+	}
+	d.cal = uint16(cal)
+
+	return nil
+}
+
+// Readings returns a list containing three items (voltage, current, and power).
+func (d *ina219) Readings(ctx context.Context, extra map[string]interface{}) (map[string]interface{}, error) {
+	handle, err := d.bus.OpenHandle(d.addr)
+	if err != nil {
+		d.logger.Errorf("can't open ina219 i2c: %s", err)
+		return nil, err
+	}
+	defer utils.UncheckedError(handle.Close())
+
+	// use the calibration result to set the scaling factor
+	// of the current and power registers for the maximum resolution
+	buf := make([]byte, 2)
+	binary.BigEndian.PutUint16(buf, d.cal)
+	err = handle.WriteBlockData(ctx, calibrationRegister, buf)
+	if err != nil {
+		return nil, err
+	}
+
+	buf = make([]byte, 2)
+	binary.BigEndian.PutUint16(buf, uint16(0x1FFF))
+	err = handle.WriteBlockData(ctx, configRegister, buf)
+	if err != nil {
+		return nil, err
+	}
+
+	var pm powerMonitor
+
+	// get shunt voltage - currently we are not returning - is it useful?
+	shunt, err := handle.ReadBlockData(ctx, shuntVoltageRegister, 2)
+	if err != nil {
+		return nil, err
+	}
+
+	// Least significant bit is 10µV.
+	pm.Shunt = int64(binary.BigEndian.Uint16(shunt)) * 10 * 1000
+	d.logger.Debugf("ina219 shunt : %d", pm.Shunt)
+
+	bus, err := handle.ReadBlockData(ctx, busVoltageRegister, 2)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check if bit zero is set, if set the ADC has overflowed.
+	if binary.BigEndian.Uint16(bus)&1 > 0 {
+		return nil, fmt.Errorf("ina219 bus voltage register overflow, register: %d", busVoltageRegister)
+	}
+
+	pm.Voltage = float64(binary.BigEndian.Uint16(bus)>>3) * 4 / 1000
```

Comment:
This calculation seems mysterious to me, but again I didn't read the data sheet. Worth explaining in a comment? Maybe not; maybe I should just RTFM.